README.rst
============
ecmwf_models
============

.. image:: https://travis-ci.org/TUW-GEO/ecmwf_models.svg?branch=master
    :target: https://travis-ci.org/TUW-GEO/ecmwf_models

.. image:: https://coveralls.io/repos/github/TUW-GEO/ecmwf_models/badge.svg?branch=master
   :target: https://coveralls.io/github/TUW-GEO/ecmwf_models?branch=master

.. image:: https://badge.fury.io/py/ecmwf-models.svg
    :target: https://badge.fury.io/py/ecmwf-models

.. image:: https://readthedocs.org/projects/ecmwf-models/badge/?version=latest
   :target: https://ecmwf-models.readthedocs.io/en/latest/

Readers and converters for data from the `ECMWF reanalysis models
<http://apps.ecmwf.int/datasets/>`_. Written in Python.

Works great in combination with `pytesmo <https://github.com/TUW-GEO/pytesmo>`_.

Citation
========

.. image:: https://zenodo.org/badge/DOI/10.5281/zenodo.593533.svg
   :target: https://doi.org/10.5281/zenodo.593533

If you use the software in a publication then please cite it using the Zenodo DOI.
Be aware that this badge links to the latest package version.

Please select your specific version at https://doi.org/10.5281/zenodo.593533 to get the DOI of that version.
You should normally always use the DOI for the specific version of your record in citations.
This is to ensure that other researchers can access the exact research artefact you used for reproducibility.

You can find additional information regarding DOI versioning at http://help.zenodo.org/#versioning

Installation
============

Install required C-libraries via conda. For installation we recommend
`Miniconda <http://conda.pydata.org/miniconda.html>`_. So please install it according
to the official installation instructions. As soon as you have the ``conda``
command in your shell you can continue:

.. code::

    conda install -c conda-forge pandas pygrib netcdf4 scipy pyresample xarray

The following command will download and install all the needed pip packages as well
as the ecmwf-model package itself.

.. code::

    pip install ecmwf_models

To create a full development environment with conda, the environment.yml file
in this repository can be used.

.. code::

    git clone git@github.com:TUW-GEO/ecmwf_models.git ecmwf_models
    cd ecmwf_models
    conda create -n ecmwf-models python=2.7 # or any other supported version
    source activate ecmwf-models
    conda env update -f environment.yml
    python setup.py develop

This script should work on Linux or OSX and uses the ``environment.yml`` file
included in this repository. On Windows the reading of grib files is not
available at the moment. On Windows a solution would be to download the ECMWF
data in netCDF format instead.

Supported Products
==================

At the moment this package supports

- **ERA Interim** (deprecated)
- **ERA5**
- **ERA5-Land** 

reanalysis data in **grib** and **netcdf** format (download, reading, time series creation) with a default spatial
sampling of 0.75 degrees (ERA Interim), 0.25 degrees (ERA5), resp. 0.1 degrees (ERA5-Land).
It should be easy to extend the package to support other ECMWF reanalysis products.
This will be done as need arises.

Contribute
==========

We are happy if you want to contribute. Please raise an issue explaining what
is missing or if you find a bug. We will also gladly accept pull requests
against our master branch for new features or bug fixes.

Development setup
-----------------

For Development we also recommend the ``conda`` environment from the
installation part.

Guidelines
----------

If you want to contribute please follow these steps:

- Fork the ecmwf_models repository to your account
- make a new feature branch from the ecmwf_models master branch
- Add your feature
- please include tests for your contributions in one of the test directories
  We use py.test so a simple function called test_my_feature is enough
- submit a pull request to our master branch


CHANGES.rst
=========
Changelog
=========

Unreleased
==========
-

Version 0.6
===========

- Add support for downloading, reading, reshuffling era5-land
- Add support for reading, reshuffling points over land only (era5 and era5-land)
- Add function to create land definition files
- Test with pinned environments

Version 0.5
===========

- Change default time steps to 6 hours.
- Add more tests, also for download functions
- Update documentation, add installation script
- Fix bugs, update command line interfaces, update dependencies
- Separate download programs for ERA5 and ERA Interim
- Change the ERA5 download api to use cdsapi instead of ecmwf api
- Update package structure to better separate between the ERA products
- Add look-up-table file for more flexibility in variable names passed by user
- Update readme

Version 0.4
===========

- Add ERA5 support (download, reading, TS conversion)
- Add netcdf support for ERA5 and ERA-Interim download (regular grid)
- Add new grid defintions: netcdf download in regular grid, grib in gaussian grid
- Add Download with spatial resampling (grib and nc) 
- Update GRIB message storing (per day instead of per message)
- Add tests for splitting downloaded files, ERA5 reading, ERA5 reshuffling, generated grids
- Add new test data

Version 0.3
===========

- Fix help text in ecmwf_repurpose command line program.
- Fix reading of metadata for variables that do not have 'levels'
- Fix wrong import when trying to read the reformatted time series data.

Version 0.2
===========

- Add reading of basic metadata fields name, depth and units.
- Fix reading of latitudes and longitudes - where flipped before.
- Fix longitude range to -180, 180.
- Add conversion to time series format.

Version 0.1
===========

- First version
- Add ERA Interim support for downloading and reading.


ecmwf_models/utils.py
# -*- coding: utf-8 -*-
# The MIT License (MIT)
#
# Copyright (c) 2019, TU Wien
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

'''
Utility functions for all data products in this package.
'''

import warnings
import os
from datetime import datetime
import xarray as xr
import pandas as pd
from datedown.fname_creator import create_dt_fpath
import argparse
import numpy as np
from netCDF4 import Dataset
from collections import OrderedDict
try:
    import pygrib
except ImportError:
    warnings.warn("pygrib has not been imported")


def str2bool(v):
    '''
    Parse a string to True/False

    Parameters
    ---------
    v : str
        String to parse, must be part of the lists below.

    Return
    ---------
    str2bool : bool
        The parsed bool from the passed string
    '''
    if v.lower() in ('yes', 'true', 't', 'y', '1'):
        return True
    elif v.lower() in ('no', 'false', 'f', 'n', '0'):
        return False
    else:
        raise argparse.ArgumentTypeError('Boolean value expected.')

def save_ncs_from_nc(input_nc, output_path, product_name,
                     filename_templ='{product}_AN_%Y%m%d_%H%M.nc'):
    """
    Split the downloaded netcdf file into daily files and add to folder structure
    necessary for reshuffling.

    Parameters
    ----------
    input_nc : str
        Filepath of the downloaded .nc file
    output_path : str
        Where to save the resulting netcdf files
    product_name : str
        Name of the ECMWF model (only for filename generation)
    filename_templ : str, optional (default: product_grid_date_time)
        Template for naming each separated nc file
    """
    localsubdirs = ['%Y', '%j']

    nc_in = xr.open_dataset(input_nc, mask_and_scale=True)

    filename_templ = filename_templ.format(product=product_name)

    for time in nc_in.time.values:
        subset = nc_in.sel(time=time)
        timestamp = pd.Timestamp(time).to_pydatetime()
        filepath = create_dt_fpath(timestamp,
                                   root=output_path,
                                   fname=filename_templ,
                                   subdirs=localsubdirs)
        if not os.path.exists(os.path.dirname(filepath)):
            os.makedirs(os.path.dirname(filepath))

        # same compression for all variables
        var_encode = {'zlib': True, 'complevel': 6}
        subset.to_netcdf(filepath, encoding={
                         var: var_encode for var in subset.variables})
    nc_in.close()

def save_gribs_from_grib(input_grib, output_path, product_name,
                         filename_templ="{product}_AN_%Y%m%d_%H%M.grb"):
    """
    Split the downloaded grib file into daily files and add to folder structure
    necessary for reshuffling.

    Parameters
    ----------
    input_grib : str
        Filepath of the downloaded .grb file
    output_path : str
        Where to save the resulting grib files
    product_name : str
        Name of the ECMWF model (only for filename generation)
    filename_templ : str, optional (default: product_OPER_0001_AN_date_time)
        Template for naming each separated grb file
    """
    localsubdirs = ['%Y', '%j']
    grib_in = pygrib.open(input_grib)

    grib_in.seek(0)
    for grb in grib_in:
        template = filename_templ
        filedate = datetime(grb['year'], grb['month'], grb['day'], grb['hour'])

        template = template.format(product=product_name)

        filepath = create_dt_fpath(filedate,
                                   root=output_path,
                                   fname=template,
                                   subdirs=localsubdirs)

        if not os.path.exists(os.path.dirname(filepath)):
            os.makedirs(os.path.dirname(filepath))

        grb_out = open(filepath, 'ab')

        grb_out.write(grb.tostring())
        grb_out.close()
    grib_in.close()

def mkdate(datestring):
    '''
    Turns a datetime string into a datetime object

    Parameters
    -----------
    datestring: str
        Input datetime string

    Returns
    -----------
    datetime : datetime
        Converted string
    '''
    if len(datestring) == 10:
        return datetime.strptime(datestring, '%Y-%m-%d')
    if len(datestring) == 16:
        return datetime.strptime(datestring, '%Y-%m-%dT%H:%M')

def parse_product(inpath):
    """
    Tries to find out what product is stored in the path. This is done based
    on the name of the first file in the path that is found.

    Parameters
    ----------
    inpath: str
        Input path where ERA data was downloaded to

    Returns
    -------
    product : str
        Product name
    """
    onedown = os.path.join(inpath, os.listdir(inpath)[0])
    twodown = os.path.join(onedown, os.listdir(onedown)[0])

    for path, subdirs, files in os.walk(twodown):
        for name in files:
            filename, extension = os.path.splitext(name)
            parts = filename.split('_')

            if 'ERA5-LAND' in parts:
                return 'era5-land'
            elif 'ERA5' in parts:
                return 'era5'
            elif 'ERAINT' in parts:
                return 'eraint'
            else:
                continue

def parse_filetype(inpath):
    """
    Tries to find out the file type by searching for
    grib or nc files two subdirectories into the passed input path.
    If function fails, grib is assumed.

    Parameters
    ----------
    inpath: str
        Input path where ERA data was downloaded to

    Returns
    -------
    filetype : str
        File type string.
    """
    onedown = os.path.join(inpath, os.listdir(inpath)[0])
    twodown = os.path.join(onedown, os.listdir(onedown)[0])

    filelist = []
    for path, subdirs, files in os.walk(twodown):
        for name in files:
            filename, extension = os.path.splitext(name)
            filelist.append(extension)

    if '.nc' in filelist and '.grb' not in filelist:
        return 'netcdf'
    elif '.grb' in filelist and '.nc' not in filelist:
        return 'grib'
    else:
        # if file type cannot be detected, guess grib
        return 'grib'

def load_var_table(name='era5', lut=False):
    '''
    Load the variables table for supported variables to download.

    Parameters
    --------
    lut : bool, optional (default: False)
        If set to true only names are loaded, so that they can be used for a LUT
        otherwise the full table is loaded
    '''
    name = name.lower()
    if name == 'era5':
        era_vars_csv = os.path.join(os.path.dirname(os.path.abspath(__file__)),
                                    'era5', 'era5_lut.csv')
    elif name == 'era5-land':
        era_vars_csv = os.path.join(os.path.dirname(os.path.abspath(__file__)),
                                    'era5', 'era5-land_lut.csv')
    elif name == 'eraint':
        era_vars_csv = os.path.join(os.path.dirname(os.path.abspath(__file__)),
                                    'erainterim', 'eraint_lut.csv')
    else:
        raise ValueError(name, 'No LUT for the selected dataset found.')

    if lut:
        dat = pd.read_csv(era_vars_csv)[['dl_name', 'long_name', 'short_name']]
    else:
        dat = pd.read_csv(era_vars_csv)

    return dat

def lookup(name, variables):
    '''
    Search the passed elements in the lookup table, if one does not exists,
    raise a Warning
    '''
    lut = load_var_table(name=name, lut=True)

    selected = []
    for var in variables:
        found = False
        for row in lut.itertuples():
            if var in row:
                selected.append(row.Index)
                found = True
                break
        if found:
            continue
        else:
            raise ValueError(
                'Passed variable {} is not in the list of supported variables.'.format(var))

    return lut.loc[selected, :]

def get_default_params(name='era5'):
    '''
    Read only lines that are marked as default variable in the csv file

    Parameters:
    ---------
    name : str
        Name of the product to get the default parameters for
    '''
    vars = load_var_table(name, lut=False)
    return vars.loc[vars.default == 1.0]


def make_era5_land_definition_file(data_file, out_file, data_file_y_res=0.25,
                                   ref_var='lsm', threshold=0.5,
                                   exclude_antarctica=True):
    """
    Create a land grid definition file from a variable within a downloaded,
    regular (netcdf) era5 file.

    Parameters:
    ---------
    data_file : str
        Path to the downloaded file that cotains the image that is used as the
        reference for creating the land definition file.
    out_file: str
        Full output path to the land defintion file to create.
    data_file_y_res : float, optional (default: 0.25)
        The resolution of the data file in latitude direction.
    ref_var: str, optional (default: 'lsm')
        A variable in the data_file that is the refernece for the land definiton.
        By default, we use the land-sea-mask variable.
    threshold: float, optional (default: 0.5)
        Threshold value below which a point is declared water, and above (or equal)
        which it is declared a land-point.
        If None is passed, then a point is declared a land point if it is not masked
        (numpy masked array) in the reference variable.
    exclude_antarctica: bool, optional (default: True)
        Cut off the definition file at -60° Lat to exclude Land Points in Antarctica.
    """
    lat_name, lon_name = 'latitude', 'longitude'
    ds_in = Dataset(data_file)
    ds_out = Dataset(out_file, 'w', format='NETCDF4')

    for dim_name in ds_in.dimensions.keys():
        ds_out.createDimension(dim_name, size=ds_in.dimensions[dim_name].size)
        ds_out.createVariable(dim_name, 'float32', (dim_name, ), zlib=True)
        ds_out.variables[dim_name][:] = ds_in.variables[dim_name][:]

    ref = ds_in.variables[ref_var]

    land_mask =  np.zeros(ref.shape)

    if np.isnan(threshold):
        land_mask[~ref[:].mask] = 1.
    else:
        land_mask[ref[:]>=threshold] = 1.

    # drop values below -60° Lat
    if exclude_antarctica:
        cut_off_lat = -60.
        index_thres_lat = ((180. / data_file_y_res) + 1) - ((90. + cut_off_lat) / data_file_y_res)
        land_mask[int(index_thres_lat):,:] = np.nan
    else:
        cut_off_lat = None


    ds_out.createVariable('land', 'float32', (lat_name, lon_name), zlib=True)
    ds_out.variables['land'][:] = land_mask

    land_attrs = OrderedDict([('units', '(0,1)'),('long_name', 'Land-sea mask'),
                              ('based_on_variable', ref_var),
                              ('standard_name', 'land_binary_mask'),
                              ('threshold_land_>=', str(threshold)),
                              ('cut_off_at', str(cut_off_lat))])

    for attr, val in land_attrs.items():
        ds_out.variables['land'].setncattr(attr, val)

    ds_in.close()
    ds_out.close()